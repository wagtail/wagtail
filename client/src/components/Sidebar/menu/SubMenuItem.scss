.sidebar-sub-menu-trigger-icon {
    display: block;
    width: 1.5em;
    height: 1.5em;
    position: absolute;
    top: 0.8125em;
    right: 0.5em;

    @include transition(transform 0.3s ease, top 0.3s ease, right 0.3s ease, width 0.3s ease, height 0.3s ease);

    &--open {
        transform-origin: 50% 50%;
        transform: rotate(180deg);
    }
}

.sidebar-sub-menu-panel {
    visibility: hidden;
    background: $nav-submenu-bg;
    transform: translate3d(0, 0, 0);
    position: fixed;
    height: 100vh;
    width: $menu-width;
    padding: 0;
    top: 0;
    left: 0;
    display: flex;
    flex-direction: column;

    @include transition(left 0.3s ease);

    > h2,
    &__list {
        width: $menu-width;
    }

    > h2 {
        display: block;
        padding: 0.2em 0;
        font-size: 1.2em;
        font-weight: 500;
        text-transform: none;
        text-align: center;
        color: $color-menu-text;

        &:before {
            font-size: 4em;
            display: block;
            text-align: center;
            margin: 0 0 0.2em;
            width: 100%;
            opacity: 0.15;
        }
    }

    > ul,
    ul > li {
        margin: 0;
        padding: 0;
        list-style-type: none;
    }

    ul > li {
        position: relative;

        @include transition(border-color 0.3s ease);
    }

    > ul {
        overflow: auto;
        flex-grow: 1;
    }

    > ul > li {
        border: 0;
    }

    &__footer {
        margin: 0;
        padding: 0.9em 1.7em;
        text-align: center;
        color: $color-menu-text;
    }

    &--visible {
        visibility: visible;
        box-shadow: 2px 0 2px rgba(0, 0, 0, 0.35);
    }

    @at-root .sidebar--slim #{&} {
        left: $menu-width-slim - $menu-width;
    }
    // Don't apply this to nested submenus though
    @at-root .sidebar--slim .sidebar-sub-menu-panel #{&} {
        left: 0;
    }

    &--open {
        left: $menu-width;

        // If another submenu is opening, display this menu behind it
        z-index: -1;

        @at-root .sidebar--slim #{&} {
            left: $menu-width-slim;
        }
        // Don't apply this to nested submenus though
        @at-root .sidebar--slim .sidebar-sub-menu-panel #{&} {
            left: $menu-width;
        }
    }
}

.sidebar-sub-menu-item {
    &--open {
        background: $nav-submenu-bg;

        > a {
            text-shadow: -1px -1px 0 rgba(0, 0, 0, 0.3);

            &:hover {
                background-color: transparent;
            }
        }
    }
}
