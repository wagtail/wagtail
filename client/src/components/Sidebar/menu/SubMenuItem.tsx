/* eslint-disable react/prop-types */

import * as React from 'react';

import Icon from '../../Icon/Icon';

import { renderMenu } from '../modules/MainMenu';
import { SidebarPanel } from '../SidebarPanel';
import { MenuItemDefinition, MenuItemProps } from './MenuItem';

interface SubMenuItemProps extends MenuItemProps<SubMenuItemDefinition> {
  slim: boolean;
}

export const SubMenuItem: React.FunctionComponent<SubMenuItemProps> = (
  { path, item, slim, state, dispatch, navigate }) => {
  const isOpen = state.navigationPath.startsWith(path);
  const isActive = isOpen || state.activePath.startsWith(path);
  const depth = path.split('.').length;
  const [isVisible, setIsVisible] = React.useState(false);

  React.useEffect(() => {
    if (isOpen) {
      // isOpen is set at the moment the user clicks the menu item
      setIsVisible(true);
    } else if (!isOpen && isVisible) {
      // When a submenu is closed, we have to wait for the close animation
      // to finish before making it invisible
      setTimeout(() => {
        setIsVisible(false);
      }, 300);
    }
  }, [isOpen]);

  const onClick = (e: React.MouseEvent) => {
    if (isOpen) {
      const pathComponents = path.split('.');
      pathComponents.pop();
      const parentPath = pathComponents.join('.');
      dispatch({
        type: 'set-navigation-path',
        path: parentPath,
      });
    } else {
      dispatch({
        type: 'set-navigation-path',
        path,
      });
    }

    e.preventDefault();
  };

  const className = (
    'sidebar-menu-item sidebar-sub-menu-item'
    + (isActive ? ' sidebar-menu-item--active' : '')
    + (isOpen ? ' sidebar-sub-menu-item--open' : '')
  );

  const sidebarTriggerIconClassName = (
    'sidebar-sub-menu-trigger-icon'
    + (isOpen ? ' sidebar-sub-menu-trigger-icon--open' : '')
  );

  return (
    <li className={className}>
      <a
        href="#"
        onClick={onClick}
        className={item.classNames}
        aria-haspopup="true"
        aria-expanded={isOpen ? 'true' : 'false'}
      >
        {item.iconName && <Icon name={item.iconName} className="icon--menuitem" />}
        <span className="menuitem-label">{item.label}</span>
        <Icon className={sidebarTriggerIconClassName} name="arrow-right" />
      </a>
      <SidebarPanel isVisible={isVisible} isOpen={isOpen} depth={depth}>
        <div className="sidebar-sub-menu-panel">
          <h2 id={`wagtail-sidebar-submenu${path.split('.').join('-')}-title`} className={item.classNames}>
            {item.iconName && <Icon name={item.iconName} className="icon--submenu-header" />}
            {item.label}
          </h2>
          <ul aria-labelledby={`wagtail-sidebar-submenu${path.split('.').join('-')}-title`}>
            {renderMenu(path, item.menuItems, slim, state, dispatch, navigate)}
          </ul>
          {item.footerText && <p className="sidebar-panel__footer">{item.footerText}</p>}
        </div>
      </SidebarPanel>
    </li>
  );
};

export class SubMenuItemDefinition implements MenuItemDefinition {
  name: string;
  label: string;
  menuItems: MenuItemDefinition[];
  iconName: string | null;
  classNames?: string;
  footerText: string;

  constructor(
    {
      name,
      label,
      icon_name: iconName = null,
      classnames = undefined,
      footer_text: footerText = '',
    }: any,
    menuItems: MenuItemDefinition[]
  ) {
    this.name = name;
    this.label = label;
    this.menuItems = menuItems;
    this.iconName = iconName;
    this.classNames = classnames;
    this.footerText = footerText;
  }

  render({ path, slim, state, dispatch, navigate }) {
    return (
      <SubMenuItem
        key={this.name}
        item={this}
        path={path}
        slim={slim}
        state={state}
        dispatch={dispatch}
        navigate={navigate}
      />
    );
  }
}
