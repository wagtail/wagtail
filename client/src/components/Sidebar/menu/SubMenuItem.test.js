import React from 'react';
import { shallow } from 'enzyme';
import { SubMenuItem } from './SubMenuItem';

describe('SubMenuItem', () => {
  const strings = {};
  const state = { activePath: '.reports.workflows', navigationPath: '' };

  it('should render with the minimum required props', () => {
    const wrapper = shallow(
      <SubMenuItem
        item={{ classNames: '', menuItems: [] }}
        items={[]}
        state={state}
        path=".reports"
        strings={strings}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should provide a button to expand the sub-menu', () => {
    const dispatch = jest.fn();

    const wrapper = shallow(
      <SubMenuItem
        dispatch={dispatch}
        item={{ classNames: '', menuItems: [] }}
        items={[]}
        state={state}
        path=".reports"
        strings={strings}
      />,
    );

    expect(wrapper.find('.sidebar-menu-item__link').prop('aria-expanded')).toBe(
      'false',
    );
    expect(dispatch).not.toHaveBeenCalled();
    expect(wrapper.find('.sidebar-sub-menu-item--open')).toHaveLength(0);

    // click the sub menu item
    wrapper.find('.sidebar-menu-item__link').simulate('click');

    // check the dispatch function (redux state) was called
    expect(dispatch).toHaveBeenCalledWith({
      path: '.reports',
      type: 'set-navigation-path',
    });

    // manually update the state as if the redux action was dispatched
    wrapper.setProps({
      state: { navigationPath: '.reports', activePath: '.reports.workflows' },
    });

    expect(wrapper.find('.sidebar-menu-item__link').prop('aria-expanded')).toBe(
      'true',
    );
    expect(wrapper.find('.sidebar-sub-menu-item--open')).toHaveLength(1);

    // click a second time to 'close'
    wrapper.find('.sidebar-menu-item__link').simulate('click');

    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(dispatch).toHaveBeenLastCalledWith({
      path: '',
      type: 'set-navigation-path',
    });
  });
});
