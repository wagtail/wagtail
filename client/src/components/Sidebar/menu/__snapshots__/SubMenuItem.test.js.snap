// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubMenuItem should render with the minimum required props 1`] = `
<li
  className="sidebar-menu-item sidebar-sub-menu-item sidebar-menu-item--active"
>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    className="sidebar-menu-item__link "
    onClick={[Function]}
    type="button"
  >
    <span
      className="menuitem-label"
    />
    <Icon
      className="sidebar-sub-menu-trigger-icon"
      name="arrow-right"
    />
  </button>
  <SidebarPanel
    depth={2}
    isOpen={false}
    isVisible={false}
  >
    <div
      className="sidebar-sub-menu-panel"
    >
      <h2
        className=""
        id="wagtail-sidebar-submenu-reports-title"
      />
      <ul
        aria-labelledby="wagtail-sidebar-submenu-reports-title"
      />
    </div>
  </SidebarPanel>
</li>
`;
